fml/platform/posix/file_posix.cc

```diff
@@ -8,46 +8,115 @@
 #include <sys/stat.h>
 #include <unistd.h>
 
+#include <memory>
+#include <sstream>
+
 #include "flutter/fml/eintr_wrapper.h"
+#include "flutter/fml/mapping.h"
 
 namespace fml {
 
-fml::UniqueFD OpenFile(const char* path,
-                       OpenPermission permission,
-                       bool is_directory) {
-  return OpenFile(fml::UniqueFD{AT_FDCWD}, path, permission, is_directory);
+std::string CreateTemporaryDirectory() {
+  char directory_name[] = "/tmp/flutter_XXXXXXXX";
+  auto result = ::mkdtemp(directory_name);
+  if (result == nullptr) {
+    return "";
+  }
+  return {result};
 }
 
-fml::UniqueFD OpenFile(const fml::UniqueFD& base_directory,
-                       const char* path,
-                       OpenPermission permission,
-                       bool is_directory) {
-  if (path == nullptr) {
-    return fml::UniqueFD{};
+static int ToPosixAccessFlags(FilePermission permission) {
+  int flags = 0;
+  switch (permission) {
+    case FilePermission::kRead:
+      flags |= O_RDONLY;  // read only
+      break;
+    case FilePermission::kWrite:
+      flags |= O_WRONLY;  // write only
+      break;
+    case FilePermission::kReadWrite:
+      flags |= O_RDWR;  // read-write
+      break;
+    case FilePermission::kExecute:
+      flags |= O_RDONLY;  // read only
+      break;
   }
+  return flags;
+}
 
-  int flags = 0;
+static int ToPosixCreateModeFlags(FilePermission permission) {
+  int mode = 0;
   switch (permission) {
-    case OpenPermission::kRead:
-      flags = O_RDONLY;
+    case FilePermission::kRead:
+      mode |= S_IRUSR;
       break;
-    case OpenPermission::kWrite:
-      flags = O_WRONLY;
+    case FilePermission::kWrite:
+      mode |= S_IWUSR;
       break;
-    case OpenPermission::kReadWrite:
-      flags = O_RDWR;
+    case FilePermission::kReadWrite:
+      mode |= S_IRUSR | S_IWUSR;
       break;
-    case OpenPermission::kExecute:
-      flags = O_RDONLY;
+    case FilePermission::kExecute:
+      mode |= 0;
       break;
   }
+  return mode;
+}
+
+fml::UniqueFD OpenFile(const char* path,
+                       bool create_if_necessary,
+                       FilePermission permission) {
+  return OpenFile(fml::UniqueFD{AT_FDCWD}, path, create_if_necessary,
+                  permission);
+}
+
+fml::UniqueFD OpenFile(const fml::UniqueFD& base_directory,
+                       const char* path,
+                       bool create_if_necessary,
+                       FilePermission permission) {
+  if (path == nullptr) {
+    return {};
+  }
 
-  if (is_directory) {
-    flags |= O_DIRECTORY;
+  int flags = 0;
+  int mode = 0;
+
+  if (create_if_necessary && !FileExists(base_directory, path)) {
+    flags = ToPosixAccessFlags(permission) | O_CREAT | O_TRUNC;
+    mode = ToPosixCreateModeFlags(permission);
```

Comment:
Would the create mode flags ever be anything other than (S_IRUSR | S_IWUSR)?

It might be simpler to eliminate ToPosixCreateModeFlags and always use read/write for newly created files and read/write/execute for newly created directories